autopush/logging.py

```diff
@@ -0,0 +1,85 @@
+# TWISTED_LOG_MESSAGE and EliotObserver licensed under APL 2.0 from
+# ClusterHQ/flocker
+# https://github.com/ClusterHQ/flocker/blob/master/flocker/common/script.py#L81-L106
+import json
+import os
+import pkg_resources
+import socket
+import sys
+
+import raven
+from eliot import add_destination, fields, Logger, MessageType
+from twisted.python.log import textFromEventDict, startLoggingWithObserver
+
+HOSTNAME = socket.getfqdn()
+LOGGER = None
+TWISTED_LOG_MESSAGE = MessageType("twisted:log",
+                                  fields(error=bool, message=unicode),
+                                  u"A log message from Twisted.")
+
+
+class EliotObserver(object):
+    """A Twisted log observer that logs to Eliot."""
+    def __init__(self):
+        if "SENTRY_DSN" in os.environ:
+            self.raven_client = raven.Client(
+                release=raven.fetch_package_version())
+        else:
+            self.raven_client = None
+        self.logger = Logger()
+
+    def raven_log(self, event):
+        f = event['failure']
+        self.raven_client.captureException(
+            (f.type, f.value, f.getTracebackObject()))
+
+    def __call__(self, msg):
+        error = bool(msg.get("isError"))
+
+        if self.raven_client and (error or 'failure' in msg):
+            self.raven_log(msg)
+            error = True
+
+        # Twisted log messages on Python 2 are bytes. We don't know the
+        # encoding, but assume it's ASCII superset. Charmap will translate
+        # ASCII correctly, and higher-bit characters just map to
+        # corresponding Unicode code points, and will never fail at decoding.
+        message = unicode(textFromEventDict(msg), "charmap")
+        kw = msg.copy()
+        for key in ["message", "isError", "failure", "why", "format"]:
+            kw.pop(key, None)
+        TWISTED_LOG_MESSAGE(error=error, message=message, **kw).write(
+            self.logger)
+
+    def start(self):
+        """Start capturing Twisted logs."""
+        startLoggingWithObserver(self, setStdout=False)
+
+
+def stdout(message):
+    msg = {}
+    ts = message.pop("timestamp")
+    del message["task_level"]
+    msg["Hostname"] = HOSTNAME
```

Comment:
Not sure, hostname is the host to bind to, but might not be the preferred name so I went with fqdn.